src/Uno.Foundation/Collections/NullableKeyDictionary.cs

```diff
@@ -0,0 +1,193 @@
+﻿#nullable enable
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Uno.Foundation.Collections
+{
+	/// <summary>
+	/// Represents a variant of dictionary that allows adding a value for a null key.
+	/// </summary>
+	/// <typeparam name="TKey">Key.</typeparam>
+	/// <typeparam name="TValue">Value.</typeparam>
+	internal class NullableKeyDictionary<TKey, TValue> : IDictionary<TKey, TValue>
+		  where TKey : class
+	{
+		private TValue _nullValue = default!;
+		private bool _containsNullValue;
+		private readonly Dictionary<TKey, TValue> _dictionary;
+
+		public NullableKeyDictionary() =>
+			_dictionary = new Dictionary<TKey, TValue>();
+
+		public NullableKeyDictionary(IEqualityComparer<TKey> comparer) =>
+			_dictionary = new Dictionary<TKey, TValue>(comparer);
+
+		public bool ContainsKey(TKey? key) => key == null ? _containsNullValue : _dictionary.ContainsKey(key);
+
+		public void Add(TKey? key, TValue value)
+		{
+			if (key == null)
+			{
+				_nullValue = value;
+				_containsNullValue = true;
+			}
+			else
+			{
+				_dictionary[key] = value;
+			}
+		}
+
+		public bool Remove(TKey? key)
+		{
+			if (key != null)
+			{
+				return _dictionary.Remove(key);
+			}
+
+			if (_containsNullValue)
+			{
+				_nullValue = default!;
+				_containsNullValue = false;
+				return true;
+			}
+
+			return false;
+		}
+
+		public bool TryGetValue(TKey? key, out TValue value)
+		{
+			if (key != null)
+			{
+				return _dictionary.TryGetValue(key, out value!);
+			}
+
+			if (_containsNullValue)
+			{
+				value = _nullValue;
+				return true;
+			}
+			else
+			{
+				value = default!;
+				return false;
+			}
+		}
+
+		public TValue this[TKey? key]
+		{
+			get
+			{
+				if (key != null)
+				{
+					_dictionary.TryGetValue(key, out var ret);
+
+					return ret!;
+				}
+				else
+				{
+					return _nullValue!;
+				}
```

Comment:
With this implementation, the return value should be nullable (when the key doesn't exist). However, the regular Dictionary throws if the key to retrieve isn't found.